File: swift/Sources/FluentCards/RefreshBuilder.swift
/// RefreshBuilder builds the refresh configuration for an Adaptive Card.
public final class RefreshBuilder {
    var data: [String: Any] = [:]

    public init() {}

    @discardableResult
    public func withAction(_ configure: (ActionBuilder) -> Void) -> RefreshBuilder {
        let ab = ActionBuilder()
        configure(ab)
        data["action"] = ab.build()
        return self
    }

    @discardableResult
    public func addUserID(_ userID: String) -> RefreshBuilder {
        if data["userIds"] == nil {
            data["userIds"] = [Any]()
        }
        var userIds = data["userIds"] as! [Any]
        userIds.append(userID)
        data["userIds"] = userIds
        return self
    }

    @discardableResult
    public func withExpires(_ expires: String) -> RefreshBuilder {
        data["expires"] = expires
        return self
    }

    public func build() -> Card {
        return data
    }
}

File: rust/fluentcards/src/builders/refresh_builder.rs
use serde_json::Value;

use crate::models::Card;
use super::action_builder::ActionBuilder;

/// Builds the refresh configuration for an Adaptive Card.
pub struct RefreshBuilder {
    data: Card,
}

impl RefreshBuilder {
    pub(crate) fn new() -> Self {
        RefreshBuilder { data: Card::new() }
    }

    pub fn with_action(
        &mut self,
        configure: impl FnOnce(&mut ActionBuilder),
    ) -> &mut Self {
        let mut ab = ActionBuilder::new();
        configure(&mut ab);
        self.data
            .insert("action".into(), Value::Object(ab.build()));
        self
    }

    pub fn add_user_id(&mut self, user_id: &str) -> &mut Self {
        let user_ids = self
            .data
            .entry("userIds")
            .or_insert_with(|| Value::Array(Vec::new()));
        if let Value::Array(arr) = user_ids {
            arr.push(Value::String(user_id.into()));
        }
        self
    }

    pub fn with_expires(&mut self, expires: &str) -> &mut Self {
        self.data
            .insert("expires".into(), Value::String(expires.into()));
        self
    }

    pub fn build(&self) -> Card {
        self.data.clone()
    }
}
